core/src/main/scala/lattice/Lattice.scala

```diff
@@ -2,38 +2,68 @@ package lattice
 
 import scala.annotation.implicitNotFound
 
+trait PartialOrderingWithBottom[V] extends PartialOrdering[V] {
+  /**
+   * Result of comparing x with operand y. Returns None if operands are not comparable. If operands are comparable, returns Some(r) where
+   * r < 0 iff x < y
+   * r == 0 iff x == y
+   * r > 0 iff x > y
+   */
+  override def tryCompare(x: V, y: V): Option[Int] =
+    if (lt(x, y)) Some(-1)
+    else if (gt(x, y)) Some(1)
+    else if (equiv(x, y)) Some(0)
+    else None
+
+  def bottom: V
+}
+
 @implicitNotFound("type ${V} does not have a Lattice instance")
-trait Lattice[V] {
+trait Lattice[V] extends PartialOrderingWithBottom[V] {
   /**
-   * Return Some(v) if a new value v was computed (v != current), else None.
-   * If it fails, throw exception.
+   * Return the join of v1 and v2 wrt. the lattice.
    */
-  def join(current: V, next: V): V
-  def empty: V
+  def join(v1: V, v2: V): V
+
+  def lteq(v1: V, v2: V): Boolean = {
+    try
+      join(v1, v2) == v2
+    catch { case _: LatticeViolationException[V] => false }
```

Comment:
I don't think we should expect an exception to be thrown by `join(v1, v2)`, because in a `Lattice[T]` the result of `join(v1, v2)` should always be well-defined. In fact, I think it should be part of the contract of trait `Lattice[T]` that its `join` method should never throw an exception. (Otherwise, we are obviously not implementing a lattice!)

Therefore, I think the `try-catch` should be removed.